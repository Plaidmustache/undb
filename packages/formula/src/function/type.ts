export type FormulaFunction =
  // 数学运算
  | "ADD"
  | "SUBTRACT"
  | "MULTIPLY"
  | "DIVIDE"
  | "SUM"
  | "CONCAT"
  | "MOD"
  | "POWER"
  | "SQRT"
  | "ABS"
  | "ROUND"
  | "FLOOR"
  | "CEILING"
  | "MIN"
  | "MAX"
  | "AVERAGE"
  // | "MEDIAN"

  // 文本处理
  | "UPPER"
  | "LOWER"
  | "TRIM"
  | "LEFT"
  | "RIGHT"
  | "MID"
  | "LEN"
  | "FIND"
  | "REPLACE"
  | "SUBSTITUTE"
  | "REPEAT"
  | "SEARCH"
  | "SUBSTR"

  // // 日期时间
  // | "NOW"
  // | "TODAY"
  // | "YEAR"
  // | "MONTH"
  // | "DAY"
  // | "HOUR"
  // | "MINUTE"
  // | "SECOND"
  // | "WEEKDAY"
  // | "DATE"

  // 逻辑运算
  | "AND"
  | "OR"
  | "NOT"
  | "IF"
  | "SWITCH"
  | "ISBLANK"
  | "ISNUMBER"
  | "ISTEXT"

  // 统计函数
  | "COUNT"
  | "COUNTA"
  | "COUNTIF"
  | "SUMIF"
  | "CORREL"
  | "RECORD_ID"
