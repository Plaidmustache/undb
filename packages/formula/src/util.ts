import { CharStreams, CommonTokenStream } from "antlr4ts"
import { FormulaLexer } from "./grammar/FormulaLexer"
import { FormulaParser } from "./grammar/FormulaParser"
import { CustomFormulaVisitor } from "./visitor"

export function createParser(input: string) {
  const inputStream = CharStreams.fromString(input)
  const lexer = new FormulaLexer(inputStream)
  const tokenStream = new CommonTokenStream(lexer)
  return new FormulaParser(tokenStream)
}

export function parseFormula(input: string) {
  const parser = createParser(input)

  const tree = parser.formula()

  const visitor = new CustomFormulaVisitor()
  const parsedFormula = visitor.visit(tree)

  return parsedFormula
}
