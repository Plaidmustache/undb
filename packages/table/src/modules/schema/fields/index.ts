export * from "./field-filter.type"
export * from "./field.type"
export * from "./field.visitor"
export * from "./variants"

export const fieldTypes = ["string", "number"] as const
