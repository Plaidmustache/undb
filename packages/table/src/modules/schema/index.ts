export * from "./dto"
export * from "./fields"
export * from "./schema.vo"
export * from "./schema.util"
