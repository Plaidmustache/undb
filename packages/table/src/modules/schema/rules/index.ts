export * from "./field-name-should-be-unique.rule"
