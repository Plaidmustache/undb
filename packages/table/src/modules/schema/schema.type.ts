import type { Field } from "./fields"

export type SchemaMap = Map<string, Field>
