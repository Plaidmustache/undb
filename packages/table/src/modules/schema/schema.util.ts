export const getNextName = (names: string[] = [], defaultName = "field", level = 0): string => {
  const name = level ? `${defaultName} (${level})` : defaultName
  return names.includes(name) ? getNextName(names, name, level + 1) : name
}
