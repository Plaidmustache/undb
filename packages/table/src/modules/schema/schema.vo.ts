import { Option, ValueObject } from "@undb/domain"
import { ZodSchema, z } from "@undb/zod"
import { objectify } from "radash"
import { WithNewFieldSpecification, WithUpdatedFieldSpecification } from "../../specifications"
import type { ICreateSchemaDTO } from "./dto"
import type { ISchemaDTO } from "./dto/schema.dto"
import {
  CreatedByField,
  FieldIdVo,
  FieldNameVo,
  IdField,
  ReferenceField,
  UpdatedAtField,
  UpdatedByField,
  type IUpdateFieldDTO,
} from "./fields"
import type { FieldId } from "./fields/field-id.vo"
import { FieldFactory } from "./fields/field.factory"
import type { Field, MutableFieldValue, NoneSystemField, SystemField } from "./fields/field.type"
import { AutoIncrementField } from "./fields/variants/autoincrement-field"
import { CreatedAtField } from "./fields/variants/created-at-field"
import type { SchemaMap } from "./schema.type"
import { getNextName } from "./schema.util"

export class Schema extends ValueObject<Field[]> {
  public fieldMapById: SchemaMap

  private constructor(public fields: Field[]) {
    super(fields)
    this.fieldMapById = new Map(this.fields.map((field) => [field.id.value, field]))
  }

  static create(dto: ICreateSchemaDTO): Schema {
    const fields = dto.map((field) => FieldFactory.create(field))
    return new Schema([
      IdField.create({ name: "id", type: "id" }),
      ...fields,
      CreatedAtField.create({ name: "createdAt", type: "createdAt" }),
      CreatedByField.create({ name: "createdBy", type: "createdBy" }),
      UpdatedAtField.create({ name: "updatedAt", type: "updatedAt" }),
      UpdatedByField.create({ name: "updatedBy", type: "updatedBy" }),
      AutoIncrementField.create({ name: "autoIncrement", type: "autoIncrement" }),
    ])
  }

  static fromJSON(dto: ISchemaDTO): Schema {
    const fields = dto.map((field) => FieldFactory.fromJSON(field))
    return new Schema(fields)
  }

  *[Symbol.iterator]() {
    yield* this.fields
  }

  $createField(field: Field): WithNewFieldSpecification {
    return new WithNewFieldSpecification(field)
  }

  createField(field: Field) {
    this.fields = [...this.fields, field]
    return this
  }

  $updateField(dto: IUpdateFieldDTO) {
    const field = this.getFieldById(new FieldIdVo(dto.id)).expect("Field not found")
    const updated = FieldFactory.fromJSON(dto)
    return new WithUpdatedFieldSpecification(field, updated)
  }

  updateField(field: Field): Schema {
    return new Schema(this.fields.map((f) => (f.id.equals(field.id) ? field : f)))
  }

  get systemFields(): SystemField[] {
    return this.fields.filter((f) => f.isSystem) as SystemField[]
  }

  get noneSystemFields(): NoneSystemField[] {
    return this.fields.filter((f) => !f.isSystem) as NoneSystemField[]
  }

  get displayFields(): Field[] {
    return this.fields.filter((f) => f.display)
  }

  get mutableFields(): Field[] {
    return this.fields.filter((f) => f.isMutable)
  }

  get mutableSchema() {
    const schema = objectify(
      this.mutableFields,
      (f) => f.id.value,
      (f) => f.valueSchema,
    )

    return z.object(schema)
  }

  get valuesSchema() {
    const schema = objectify(
      this.fields,
      (f) => f.id.value,
      (f) => f.valueSchema,
    )

    return z.object(schema)
  }

  get readableSchema() {
    const schema = objectify(
      this.fields,
      (f) => f.name.value,
      (f) => f.valueSchema,
    )

    return z.object(schema)
  }

  get mutateSchema() {
    const schema = objectify(
      this.mutableFields,
      (f) => f.name.value,
      (f) => (f.mutateSchema as Option<ZodSchema>).unwrap(),
    )

    return z.object(schema)
  }

  get searchableFields(): Field[] {
    return this.fields.filter((f) => f.searchable)
  }

  getNextFieldName(defaultName?: string): string {
    return new FieldNameVo(
      getNextName(
        this.fields.map((f) => f.name.value),
        defaultName,
      ),
    ).value
  }

  toJSON(): ISchemaDTO {
    return this.fields.map((field) => field.toJSON())
  }

  getFieldById(fieldId: FieldId): Option<Field> {
    const field = this.fieldMapById.get(fieldId.value)
    return Option(field)
  }

  getDefaultValues() {
    const values: Record<string, any> = {}

    for (const field of this.mutableFields) {
      if (!field) continue
      if (!field.isDefaultValueValid) continue
      const defaultValue = field.defaultValue as Option<MutableFieldValue>
      if (defaultValue.isNone()) continue

      values[field.id.value] = defaultValue.unwrap().value
    }

    return values
  }

  getDisplayFields(fieldIds?: Set<string>) {
    return this.fields.filter((f) => (fieldIds ? fieldIds.has(f.id.value) : true) && f.display)
  }

  getReferenceFields(fields: Field[] = this.fields): ReferenceField[] {
    return fields.filter((f) => f instanceof ReferenceField) as ReferenceField[]
  }

  getForeignTableIds(fields: Field[] = this.fields): Set<string> {
    const referenceFields = this.getReferenceFields(fields)

    return new Set(referenceFields.map((f) => f.foreignTableId))
  }
}
