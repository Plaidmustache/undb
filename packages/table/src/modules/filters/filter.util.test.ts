import { describe, expect, test } from 'bun:test'
import { Schema } from '../schema'
import type { IFilterGroup, MaybeFilterGroup } from './filter.type'
import { getSpec, parseValidFilter } from './filter.util'

const schema = Schema.fromJSON([
  { id: 'fld_1', type: 'string', name: 'fld_1' },
  { id: 'fld_2', type: 'number', name: 'fld_2' },
])

describe('filter.util', () => {
  test.each<IFilterGroup>([
    {
      conjunction: 'and',
      children: [
        { fieldId: 'fld_1', op: 'eq', value: 'value1' },
        { fieldId: 'fld_2', op: 'gt', value: 1 },
      ],
    },
    {
      conjunction: 'or',
      children: [
        { fieldId: 'fld_1', op: 'eq', value: 'value1' },
        {
          conjunction: 'and',
          children: [
            { fieldId: 'fld_1', op: 'eq', value: 'value1' },
            { fieldId: 'fld_2', op: 'gt', value: 1 },
          ],
        },
        {
          conjunction: 'or',
          children: [
            { fieldId: 'fld_1', op: 'eq', value: 'value2' },
            { fieldId: 'fld_2', op: 'lt', value: 2 },
          ],
        },
      ],
    },
  ])('should get correct spec', (filter) => {
    const spec = getSpec(schema.fieldMapById, filter)
    expect(spec).toMatchSnapshot()
  })

  describe('parseValidFilter', () => {
    test.each<IFilterGroup>([
      {
        conjunction: 'and',
        children: [
          { fieldId: 'fld_1', op: 'eq', value: 'value1' },
          { fieldId: 'fld_2', op: 'gt', value: 1 },
        ],
      },
      {
        conjunction: 'or',
        children: [
          { fieldId: 'fld_1', op: 'eq', value: 'value1' },
          {
            conjunction: 'and',
            children: [
              { fieldId: 'fld_1', op: 'eq', value: 'value1' },
              { fieldId: 'fld_2', op: 'gt', value: 1 },
            ],
          },
          {
            conjunction: 'or',
            children: [
              { fieldId: 'fld_1', op: 'eq', value: 'value2' },
              { fieldId: 'fld_2', op: 'lt', value: 2 },
            ],
          },
        ],
      },
    ])('should parse valid filter', (filter) => {
      const parsed = parseValidFilter(schema.fieldMapById, filter)
      expect(parsed).toEqual(filter)
    })

    test.each<[MaybeFilterGroup, IFilterGroup]>([
      [
        {
          conjunction: 'and',
          children: [
            { fieldId: 'fld_1', op: 'eq', value: 'value1' },
            { fieldId: 'fld_2', op: 'gt', value: '1' },
          ],
        },
        {
          conjunction: 'and',
          children: [{ fieldId: 'fld_1', op: 'eq', value: 'value1' }],
        },
      ],
      [
        {
          conjunction: 'or',
          children: [
            { fieldId: 'fld_1', op: 'eq', value: 'value1' },
            {
              conjunction: 'and',
              children: [
                { fieldId: 'fld_1', op: 'eq', value: 'value1' },
                { fieldId: 'fld_2', op: 'gt', value: '1' },
              ],
            },
            {
              conjunction: 'or',
              children: [
                { fieldId: 'fld_1', value: 'value2' },
                { fieldId: 'fld_2', op: 'lt', value: 2 },
              ],
            },
          ],
        },
        {
          conjunction: 'or',
          children: [
            { fieldId: 'fld_1', op: 'eq', value: 'value1' },
            {
              conjunction: 'and',
              children: [{ fieldId: 'fld_1', op: 'eq', value: 'value1' }],
            },
            {
              conjunction: 'or',
              children: [{ fieldId: 'fld_2', op: 'lt', value: 2 }],
            },
          ],
        },
      ],
    ])('should ignore invalid filter', (filter, value) => {
      const parsed = parseValidFilter(schema.fieldMapById, filter)
      expect(parsed).toEqual(value)
    })
  })
})
