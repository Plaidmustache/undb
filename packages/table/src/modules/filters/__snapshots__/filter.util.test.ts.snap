// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`filter.util should get correct spec 1`] = `
OptionType {
  [Symbol(T)]: true,
  [Symbol(Val)]: And {
    "left": StringEqual {
      "fieldId": {
        "props": {
          "value": "fld_1",
        },
      },
      "values": StringFieldValue {
        "props": {
          "value": "value1",
        },
      },
    },
    "right": NumberGT {
      "fieldId": {
        "props": {
          "value": "fld_2",
        },
      },
      "value": 1,
    },
  },
}
`;

exports[`filter.util should get correct spec 2`] = `
OptionType {
  [Symbol(T)]: true,
  [Symbol(Val)]: Or {
    "left": Or {
      "left": StringEqual {
        "fieldId": {
          "props": {
            "value": "fld_1",
          },
        },
        "values": StringFieldValue {
          "props": {
            "value": "value1",
          },
        },
      },
      "right": And {
        "left": StringEqual {
          "fieldId": {
            "props": {
              "value": "fld_1",
            },
          },
          "values": StringFieldValue {
            "props": {
              "value": "value1",
            },
          },
        },
        "right": NumberGT {
          "fieldId": {
            "props": {
              "value": "fld_2",
            },
          },
          "value": 1,
        },
      },
    },
    "right": Or {
      "left": StringEqual {
        "fieldId": {
          "props": {
            "value": "fld_1",
          },
        },
        "values": StringFieldValue {
          "props": {
            "value": "value2",
          },
        },
      },
      "right": NumberLT {
        "fieldId": {
          "props": {
            "value": "fld_2",
          },
        },
        "value": 2,
      },
    },
  },
}
`;
